### PR TITLE
Tool 1100 yarn cache

### PR DESCRIPTION
**Pull request by lpusok**
Added go implementation, caching for node_modules.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+func main() {
+	var config config
+	if err := stepconf.Parse(&config); err != nil {
+		failf("Issue with input: %s", err)
+	}
+	stepconf.Print(config)
+	fmt.Println()
+	log.SetEnableDebugLog(config.IsDebugLog == "yes")
```
Comment (BirmacherAkos) [inline main.go]: `IsDebugLog` is a bool, you don't need to check it as a string

File: main.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+	if path, err := exec.LookPath("yarn"); err != nil {
+		log.Infof("Yarn not installed. Installing...")
+		installCmd, err := getInstallYarnCommand()
+		if err != nil {
+			failf("%s", err)
```
Comment (BirmacherAkos) [inline main.go]: Write please what command call caused the error.
E.g: failf("Failed to create yarn install command", err)

File: main.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+	log.Infof("Yarn version:")
+	versionCmd := command.New("yarn", "--version")
+	versionCmd.SetStdout(os.Stdout).SetStderr(os.Stderr).SetDir(absWorkingDir)
+
+	fmt.Println()
+	log.Donef("$ %s", versionCmd.PrintableCommandArgs())
+	fmt.Println()
+	err = versionCmd.Run()
+	if err != nil {
```
Comment (BirmacherAkos) [inline main.go]: You could do this one line

```suggestion
	if err = versionCmd.Run(); err != nil {
```

File: bitrise.yml [hunk 4]
```diff
@@ -1,16 +1,26 @@
-format_version: 1.0.0
+format_version: 8
```
Comment (BirmacherAkos) [inline bitrise.yml]: Note:
The current format version is 9. But 8 is perfectly fine here.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	"github.com/bitrise-io/go-utils/log"
+	"github.com/kballard/go-shellquote"
+)
+
+type config struct {
+	WorkingDir  string `env:"workdir,dir"`
+	YarnCommand string `env:"command"`
+	YarnArgs    string `env:"args"`
+	UseCache    string `env:"cache_local_deps,opt[yes,no]"`
+	IsDebugLog  string `env:"verbose_log,opt[yes,no]"`
```
Comment (BirmacherAkos) [inline main.go]: These two variables should be bool instead.

File: main.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+func getInstallYarnCommand() (*command.Model, error) {
+	if runtime.GOOS != "linux" {
+		return nil, fmt.Errorf("Unsupported platform %s, failed to install yarn", runtime.GOOS)
+	}
+	if _, err := os.Stat(path.Join("etc", "lsb-release")); err != nil {
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("only Ubuntu distribution supported, failed to install yarn")
+		}
+		return nil, err
+	}
```
Comment (BirmacherAkos) [inline main.go]: It will not install yarn on OSX?

File: main.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	if err != nil {
+		if errorutil.IsExitStatusError(err) {
+			failf("yarn version command failed, error: %s", err)
+		}
+		failf("Failed to run command, error: %s", err)
+	}
+
+	commandParams, err := shellquote.Split(config.YarnCommand)
+	if err != nil {
+		failf("failed to split command arguments, error: %s", err)
```
Comment (BirmacherAkos) [inline main.go]: Start every `failf` message with a capital letter.

File: main.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+	}()
+
+	if err := yarnCmd.Run(); err != nil {
+		if errorutil.IsExitStatusError(err) {
+			failf("yarn command failed, error: %s", err)
+		}
+		failf("failed to run yarn command, error: %s", err)
+	}
+
+	if config.UseCache == "yes" && (len(commandParams) == 0 || commandParams[0] == "install") {
```
Comment (BirmacherAkos) [inline main.go]: `config.UseCache` should be a bool

File: main.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+		if err != nil {
+			return err
+		}
+		if fileInfo.IsDir() && fileInfo.Name() == "node_modules" {
+			cachePaths = append(cachePaths, path)
+			return filepath.SkipDir
+		}
+		return nil
+	}); err != nil {
+		failf("Failed to find node_modules directories, error: %s", err)
```
Comment (BirmacherAkos) [inline main.go]: - You should not use `failf` in a method, just return the error and use `failf` in the `main()`.
- I think the step should not fail if the cache failed. I would write a warning only. WDYT? 🤔

**Review by BirmacherAkos — CHANGES_REQUESTED** [review]
Please check my comments

File: main.go (hunk 6 above)
Comment (lpusok) [reply main.go]: Not from here, but from the step.yml through the brew dep.

File: main.go (hunk 9 above)
Comment (lpusok) [reply main.go]: Agree with the warning, changed the code.

**Review by BirmacherAkos — APPROVED** [review]

**Review by BirmacherAkos — CHANGES_REQUESTED** [review]
Do not forget the CI

**Review by BirmacherAkos — DISMISSED** [review]

**Comment by lpusok** [comment]
@trapacska please merge the PR.

**Review by BirmacherAkos — DISMISSED** [review]

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,19 @@
+FROM golang:1.12
```
Comment (trapacska) [inline Dockerfile:1]: you should use our base image

**Review by trapacska — DISMISSED** [review]

**Review by trapacska — APPROVED** [review]

**Pull request merged**